File: csharp/SwimmingCompetitionCS/persistence/ChallengeDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using model;

namespace persistence
{   
    public class ChallengeDBRepository : ChallengeRepository
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(ChallengeDBRepository));

        IDictionary<String, string> props;

        public ChallengeDBRepository(IDictionary<string, string> props)
        {
            this.props = props;
        }

        public Challenge FindOne(long along)
        {  
            log.InfoFormat("Finding a challenge with id {0}", along);
            IDbConnection connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from challenge where id = @id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = along;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        Style style = (Style)Enum.Parse(typeof(Style), dataR.GetString(1));
                        Distance distance = (Distance)dataR.GetInt16(2);
                        Challenge challenge = new Challenge(distance, style);
                        challenge.Id = id;
                        return challenge;
                    }
                }
            }
            log.InfoFormat("Finding NO challenge with id {0}", along);
            return null;
        }

        public IEnumerable<Challenge> FindAll()
        {
            log.InfoFormat("Finding all challenges");
            IDbConnection connection = DBUtils.getConnection(props);
            IList<Challenge> challenges = new List<Challenge>();
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "select * from challenge";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        long id = dataR.GetInt32(0);
                        Style style = (Style)Enum.Parse(typeof(Style), dataR.GetString(1));
                        Distance distance = (Distance)dataR.GetInt16(2);
                        Challenge challenge = new Challenge(distance, style);
                        challenge.Id = id;
                        challenges.Add(challenge);
                    }
                }
            }
            log.Info(challenges);
            return challenges;
        }

        public Challenge Save(Challenge entity)
        {
            log.InfoFormat("Save a challenge {0}", entity);
            var connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "insert into challenge(distance, style) values(@distance,@style) RETURNING id";
                var paramDistance = comm.CreateParameter();
                paramDistance.ParameterName = "@distance";
                paramDistance.Value = (int)entity.Distance;
                comm.Parameters.Add(paramDistance);
                
                var paramStyle = comm.CreateParameter();
                paramStyle.ParameterName = "@style";
                paramStyle.Value = entity.Style.ToString();
                comm.Parameters.Add(paramStyle);

                long generatedId = (long)comm.ExecuteScalar();
                entity.Id = generatedId;

                return entity;
            }
        }

        public void Delete(long id)
        {
            log.InfoFormat("Delete challenge with id {0}", id);
            IDbConnection connection = DBUtils.getConnection(props);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "delete from challenge where id = @id";
                var idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                comm.Parameters.Add(idParam);
                var result = comm.ExecuteNonQuery();
                if(result == 1)
                    log.InfoFormat("Deleted challenge with id {0}", id);
            }
            
        }

        public void Update(Challenge entity)
        {
            // nu e nevoie
        }
    }
}

File: java/SwimmingCompetition/SwimmingCompetitionPersistence/src/main/java/swimmingcompetition/persistence/repository/ChallengeDBRepository.java
package swimmingcompetition.persistence.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import swimmingcompetition.model.Challenge;
import swimmingcompetition.model.Distance;
import swimmingcompetition.model.Style;
import swimmingcompetition.persistence.ChallengeRepository;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ChallengeDBRepository implements ChallengeRepository {
    private JdbcUtils dbUtils;
    private static final Logger logger= LogManager.getLogger(ChallengeDBRepository.class);
    public ChallengeDBRepository(Properties props) {
        this.dbUtils = new JdbcUtils(props);
    }

    @Override
    public Challenge findOne(Long aLong) {
        logger.traceEntry("finding challenge with id {}", aLong);
        Connection connection = this.dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from challenge where id = ?")){
            preparedStatement.setLong(1, aLong);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    Long id = resultSet.getLong("id");
                    Style style = Style.valueOf(resultSet.getString("style"));
                    Distance distance = Distance.fromInt(resultSet.getInt("distance"));
                    Challenge challenge = new Challenge(distance, style);
                    challenge.setId(id);
                    return challenge;
                }
            }
        }
        catch (SQLException ex){
            logger.error(ex);
            System.out.println("Eroare la DB -- Challenge " + ex.getMessage().toString());
        }
        logger.traceExit("No challenge found with id {}", aLong);
        return null;
    }

    @Override
    public Iterable<Challenge> findAll() {
        logger.traceEntry("find all challenges");
        Connection connection = dbUtils.getConnection();
        List<Challenge> challengeList = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from challenge")) {
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Long id = resultSet.getLong("id");
                    Distance distance = Distance.fromInt(resultSet.getInt("distance"));
                    Style style = Style.valueOf(resultSet.getString("style"));
                    Challenge challenge = new Challenge(distance, style);
                    challenge.setId(id);
                    challengeList.add(challenge);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Challenge " + e.getMessage().toString());
        }
        logger.traceExit(challengeList);
        return challengeList;

    }

    @Override
    public Challenge save(Challenge entity) {
        logger.traceEntry("save challenge {}", entity);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into challenge(distance, style) values(?,?)", Statement.RETURN_GENERATED_KEYS)) {
            preparedStatement.setInt(1, entity.getDistance().get());
            preparedStatement.setString(2, entity.getStyle().name());
            int result = preparedStatement.executeUpdate();
            ResultSet  resultSet = preparedStatement.getGeneratedKeys();
            if(resultSet.next()){
                Long id = resultSet.getLong(1);
                entity.setId(id);
            }
            logger.trace("saved challenge {}", entity);
            return entity;
        }
        catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Challenge " + e.getMessage().toString());
        }
        return null;
    }

    @Override
    public void delete(Long aLong) {
        logger.traceEntry("delete challenge with id {}", aLong);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from challenge where id = ?")) {
            preparedStatement.setLong(1, aLong);
            int result = preparedStatement.executeUpdate();
            logger.trace("challenges deleted {}", result);
        }
        catch (SQLException e) {
            logger.error(e);
            System.out.println("Error la DB -- Challenge " + e.getMessage().toString());
        }
    }

    @Override
    public void update(Challenge entity) {
        // Nu cred ca avem nevoie
    }
}
